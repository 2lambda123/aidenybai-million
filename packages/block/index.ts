export { createBlock, Block, stringToDOM } from './block';
export { fragment, FragmentBlock } from './fragment';
export { renderToTemplate } from './template';
export { AbstractBlock, Hole, type Props } from './types';
