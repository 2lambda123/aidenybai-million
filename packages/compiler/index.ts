import { unplugin } from './plugin';
import type { Options } from './options';

export const vite = unplugin.vite;
export const webpack = unplugin.webpack;
export const rollup = unplugin.rollup;
export const rspack = unplugin.rspack;
export const esbuild = unplugin.esbuild;
export const next = (
  nextConfig: Record<string, any> = {},
  overrideOptions: Options = {}
) => {
  const millionConfig: Options = {
    mode: 'react',
    server: true,
    ...overrideOptions,
  };
  return {
    ...nextConfig,
    webpack(config: Record<string, any>, webpackOptions: Record<string, any>) {
      config.plugins.unshift(
        webpack({
          mute: webpackOptions.isServer,
          ...millionConfig,
        })
      );

      if (typeof nextConfig.webpack === 'function') {
        return nextConfig.webpack(config, webpackOptions);
      }
      return config;
    },
  };
};

export default {
  vite,
  webpack,
  rollup,
  rspack,
  esbuild,
  next,
  unplugin,
};
