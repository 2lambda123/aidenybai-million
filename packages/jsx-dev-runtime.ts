import type { Props, VNode } from './million';

export const jsxDEV = (
  type: string,
  props: Props | null = {},
  ...children: VNode[]
): VNode => {
  if (props === null) props = {};
  props.children = children;
  return {
    type,
    props,
  };
};
export { jsxDEV as createElement, jsxDEV as jsx, jsxDEV as jsxs };
