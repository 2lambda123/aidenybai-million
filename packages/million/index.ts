export { createBlock, Block, stringToDOM, withKey } from './block';
export { fragment, FragmentBlock } from './fragment';
export { renderToTemplate } from './template';
export { AbstractBlock, type Props, type VNode } from './types';
