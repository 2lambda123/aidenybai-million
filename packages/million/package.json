{
  "type": "module",
  "private": true,
  "main": "./index.ts",
  "module": "./index.ts",
  "devDependencies": {
    "@types/react": "^18.2.46"
  }
}
