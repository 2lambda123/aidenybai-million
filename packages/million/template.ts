import {
  firstChild$,
  mapGet$,
  mapHas$,
  mapSet$,
  nextSibling$,
  setHas$,
} from './dom';
import { X_CHAR, VOID_ELEMENTS } from './constants';
import { AbstractBlock, Flags } from './types';
import type {
  Edit,
  VNode,
  InitChild,
  InitEvent,
  EditEvent,
  EditStyleAttribute,
  EditSvgAttribute,
  EditAttribute,
  EditChild,
  InitBlock,
} from './types';

export const renderToTemplate = (
  vnode: VNode,
  edits: Edit[] = [],
  path: number[] = [],
): string => {
  if (typeof vnode === 'string') {
    return vnode;
  }
  if (
    typeof vnode === 'number' ||
    typeof vnode === 'bigint' ||
    vnode === true
  ) {
    return String(vnode);
  }
  if (vnode === null || vnode === undefined || vnode === false) {
    return '';
  }

  let props = '';
  let children = '';

  const current: Edit = {
    edits: {},
    inits: null,
    getRoot: (
      root: HTMLElement,
      cache: Map<number, HTMLElement> | null,
      key: number | null,
    ) => {
      if (cache && key !== null && mapHas$.call(cache, path)) {
        return mapGet$.call(cache, path)!;
      }
      const el = getCurrentElement(path, root);
      if (cache && key !== null) mapSet$.call(cache, key, el);
      return el;
    },
  };
  const editsPointer: Edit['edits'] = {};
  let initsPointer = current.inits;

  for (let name in vnode.props) {
    const value = vnode.props[name];
    if (name === 'key' || name === 'ref' || name === 'children') {
      continue;
    }
    if (name === 'className') name = 'class';
    if (name === 'htmlFor') name = 'for';

    if (name.startsWith('on')) {
      const isValueHole = '__key' in value;
      const hole = value.__key;
      // Make edits monomorphic
      if (isValueHole && hole !== undefined && hole !== null) {
        const next: EditEvent = {
          type: Flags.EVENT,
          name,
          value: null,
          hole,
          index: null,
          listener: null,
          patch: null,
          block: null,
          next: null,
        };
        mountNode(current.edits, hole, next);
        editsPointer[hole] = addNode(editsPointer[hole], next);
      } else {
        const next: InitEvent = {
          type: Flags.EVENT,
          name,
          value: null,
          hole: null,
          index: null,
          listener: value,
          patch: null,
          block: null,
          next: null,
        };
        mountNode(current, 'inits', next);
        initsPointer = addNode(initsPointer, next);
      }

      continue;
    }

    if (value) {
      if (typeof value === 'object' && '__key' in value) {
        const hole = value.__key;
        const next: EditAttribute | EditStyleAttribute | EditSvgAttribute = {
          type:
            name === 'style'
              ? Flags.STYLE_ATTRIBUTE
              : name.charCodeAt(0) === X_CHAR
              ? Flags.SVG_ATTRIBUTE
              : Flags.ATTRIBUTE,
          name,
          value: null,
          hole,
          index: null,
          listener: null,
          patch: null,
          block: null,
          next: null,
        };

        mountNode(current.edits, hole, next);
        editsPointer[hole] = addNode(editsPointer[hole], next);
      } else {
        if (name === 'style') {
          let style = '';
          for (const key in value) {
            style += `${key}:${String(value[key])};`;
          }
          props += ` style="${style}"`;
        }
        props += ` ${name}="${String(value)}"`;
      }
    }
  }

  if (setHas$.call(VOID_ELEMENTS, vnode.type)) {
    return `<${vnode.type}${props} />`;
  }

  // 👎: 'foo' + Block + 'bar' => 'foobaz'.
  //                                      ↕️ Block edit here
  // 👍: 'foo' + Block + 'bar'   => 'foo', 'bar'
  let canMergeString = false;
  for (let i = 0, j = vnode.props.children?.length || 0, k = 0; i < j; ++i) {
    const child = vnode.props.children?.[i];
    if (child === null || child === undefined || child === false) continue;

    if (typeof child === 'object' && '__key' in child) {
      const hole = child.__key;
      const next: EditChild = {
        type: Flags.CHILD,
        name: null,
        value: null,
        hole,
        index: i,
        listener: null,
        patch: null,
        block: null,
        next: null,
      };

      mountNode(current.edits, hole, next);
      editsPointer[hole] = addNode(editsPointer[hole], next);
      continue;
    }

    if (child instanceof AbstractBlock) {
      const next: InitBlock = {
        type: Flags.BLOCK,
        name: null,
        value: null,
        hole: null,
        index: i,
        listener: null,
        patch: null,
        block: child,
        next: null,
      };

      mountNode(current, 'inits', next);
      initsPointer = addNode(initsPointer, next);
      continue;
    }

    if (
      typeof child === 'string' ||
      typeof child === 'number' ||
      typeof child === 'bigint'
    ) {
      const value =
        typeof child === 'number' || typeof child === 'bigint'
          ? String(child)
          : child;
      if (canMergeString) {
        const next: InitChild = {
          type: Flags.CHILD,
          name: null,
          value,
          hole: null,
          index: i,
          listener: null,
          patch: null,
          block: null,
          next: null,
        };
        initsPointer = addNode(initsPointer, next);
        continue;
      }
      canMergeString = true;
      children += value;
      k++;
      continue;
    }

    canMergeString = false;
    children += renderToTemplate(child, edits, [...path, k++]);
  }

  if (current.inits || !isObjectEmpty(current.edits)) {
    edits.push(current);
  }

  return `<${vnode.type}${props}>${children}</${vnode.type}>`;
};

const getCurrentElement = (path: number[], root: HTMLElement): HTMLElement => {
  const pathLength = path.length;
  if (!pathLength) return root;
  // path is an array of indices to traverse the DOM tree
  // For example, [0, 1, 2] becomes:
  // root.firstChild.firstChild.nextSibling.firstChild.nextSibling.nextSibling
  // We use path because we don't have the actual DOM nodes until mount()
  for (let i = 0; i < pathLength; ++i) {
    const siblings = path[i]!;
    // https://www.measurethat.net/Benchmarks/Show/15652/0/childnodes-vs-children-vs-firstchildnextsibling-vs-firs
    root = firstChild$.call(root);
    if (!siblings) continue;
    for (let j = 0; j < siblings; ++j) {
      root = nextSibling$.call(root) as HTMLElement;
    }
  }
  return root;
};

const isObjectEmpty = (object: Record<string, unknown>) => {
  // eslint-disable-next-line no-unreachable-loop
  for (const _ in object) return false;
  return true;
};

const mountNode = (root: any, property: string, node: any) => {
  if (!root[property]) root[property] = node;
  return node;
};

const addNode = (pointer: any | null, next: any) => {
  if (pointer) {
    pointer.next = next;
    return next;
  }
  return next;
};
