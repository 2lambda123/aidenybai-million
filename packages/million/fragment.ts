/* eslint-disable @typescript-eslint/unbound-method */
import { mapGet$, mapSet$, setTextContent$ } from './dom';
import { AbstractBlock } from './types';
import { mount$, patch$, move$, remove$ } from './block';

export const fragment = (children: AbstractBlock[], hints?: number[]) => {
  return new FragmentBlock(children, hints);
};

export class FragmentBlock extends AbstractBlock {
  b: AbstractBlock[];
  h?: number[];
  constructor(children: AbstractBlock[], hints?: number[]) {
    super();
    this.b = children;
    this.h = hints;
  }
  v() {
    /**/
  }
  // @ts-expect-error override
  override p(fragment: FragmentBlock) {
    const oldChildren = this.b;
    const newChildren = fragment.b;
    const oldChildrenLength = oldChildren.length;
    const newChildrenLength = newChildren.length;
    const parent = this.t()!;
    if (this === fragment) return;
    if (newChildrenLength === 0 && oldChildrenLength === 0) return;
    if (fragment.h) {
      const hints = fragment.h;
      for (let i = 0, j = hints.length; i < j; ++i) {
        const hint = hints[i]!;
        const oldChild: AbstractBlock | undefined = oldChildren[hint];
        const newChild: AbstractBlock | undefined = newChildren[hint];
        if (oldChild && newChild) {
          patch$.call(oldChild, newChild);
        } else if (oldChild) {
          remove$.call(oldChild.l);
        } else if (newChild) {
          // FIXME: Maybe wrong refNode
          mount$.call(newChild, parent, newChild.l);
        }
      }
      return;
    }

    this.b = newChildren;

    if (newChildrenLength === 0) {
      fragmentRemove$.call(this);
      return;
    }
    if (oldChildrenLength === 0) {
      fragmentMount$.call(fragment, parent);
      return;
    }

    let oldHead = 0;
    let newHead = 0;
    let oldTail = oldChildrenLength - 1;
    let newTail = newChildrenLength - 1;

    let oldHeadChild = oldChildren[0];
    let newHeadChild = newChildren[0]!;
    let oldTailChild = oldChildren[oldTail];
    let newTailChild = newChildren[newTail]!;

    let oldKeyMap: Map<string, number> | undefined;

    while (oldHead <= oldTail && newHead <= newTail) {
      if (!oldHeadChild) {
        oldHeadChild = oldChildren[++oldHead];
        continue;
      }
      if (!oldTailChild) {
        oldTailChild = oldChildren[--oldTail];
        continue;
      }

      const oldHeadKey = oldHeadChild.k!;
      const newHeadKey = newHeadChild.k!;
      if (oldHeadKey === newHeadKey) {
        patch$.call(oldHeadChild, newHeadChild);
        newChildren[newHead] = oldHeadChild;
        oldHeadChild = oldChildren[++oldHead];
        newHeadChild = newChildren[++newHead]!;
        continue;
      }

      const oldTailKey = oldTailChild.k!;
      const newTailKey = newTailChild.k!;
      if (oldTailKey === newTailKey) {
        patch$.call(oldTailChild, newTailChild);
        newChildren[newTail] = oldTailChild;
        oldTailChild = oldChildren[--oldTail];
        newTailChild = newChildren[--newTail]!;
        continue;
      }

      if (oldHeadKey === newTailKey) {
        patch$.call(oldHeadChild, newTailChild);
        newChildren[newTail] = oldHeadChild;
        const nextChild = newChildren[newTail + 1];
        move$.call(oldHeadChild, nextChild, nextChild?.l || null);
        oldHeadChild = oldChildren[++oldHead];
        newTailChild = newChildren[--newTail]!;
        continue;
      }

      if (oldTailKey === newHeadKey) {
        patch$.call(oldTailChild, newHeadChild);
        newChildren[newHead] = oldTailChild;
        const nextChild = oldChildren[oldHead];
        move$.call(oldTailChild, nextChild, nextChild?.l || null);
        oldTailChild = oldChildren[--oldTail];
        newHeadChild = newChildren[++newHead]!;
        continue;
      }

      if (!oldKeyMap) {
        oldKeyMap = new Map<string, number>();
        for (let i = oldHead; i <= oldTail; i++) {
          mapSet$.call(oldKeyMap, oldChildren[i]!.k!, i);
        }
      }
      const oldIndex = mapGet$.call(oldKeyMap, newHeadKey);
      if (oldIndex === undefined) {
        mount$.call(newHeadChild, parent, oldHeadChild.l || null);
      } else {
        const oldChild = oldChildren[oldIndex]!;
        move$.call(oldChild, oldHeadChild, null);
        patch$.call(oldChild, newHeadChild);
        newChildren[newHead] = oldChild;
        oldChildren[oldIndex] = null as any;
      }
      newHeadChild = newChildren[++newHead]!;
    }

    if (oldHead <= oldTail || newHead <= newTail) {
      if (oldHead > oldTail) {
        const nextChild = newChildren[newTail + 1];
        for (let i = newHead; i <= newTail; i++) {
          mount$.call(newChildren[i], parent, nextChild ? nextChild.l : null);
        }
      } else {
        for (let i = oldHead; i <= oldTail; ++i) {
          remove$.call(oldChildren[i]);
        }
      }
    }
    return this._t;
  }
  m(parent: HTMLElement, refNode: Node | null = null): HTMLElement {
    if (this._t) return this._t;
    for (let i = 0, j = this.b.length; i < j; ++i) {
      const block = this.b[i]!;
      mount$.call(block, parent, refNode);
    }
    this._t = parent;
    return parent;
  }
  x() {
    const parent = this.t();
    if (parent) {
      setTextContent$.call(parent, '');
    } else {
      for (let i = 0, j = this.b.length; i < j; ++i) {
        remove$.call(this.b[i]!);
      }
    }
    this.b = [];
  }
  u(): boolean {
    return true;
  }
  s() {
    return this.b.map((block) => block.s()).join('');
  }
  t(): HTMLElement | null | undefined {
    if (!this._t) this._t = this.b[0]!.t();
    return this._t;
  }
}

export const fragmentMount$ = FragmentBlock.prototype.m;
export const fragmentPatch$ = FragmentBlock.prototype.p;
export const fragmentRemove$ = FragmentBlock.prototype.x;
