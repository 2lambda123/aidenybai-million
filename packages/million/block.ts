/* eslint-disable no-bitwise */
/* eslint-disable @typescript-eslint/unbound-method */
import type { MillionProps } from '../types';
import {
  cloneNode$,
  createEventListener,
  insertBefore$,
  insertText,
  remove$ as removeElement$,
  setAttribute,
  setText,
  setStyleAttribute,
  setSvgAttribute,
  childAt,
  replaceChild$,
  stringToDOM,
  removeComments,
  appendChild$,
  betweenNodes,
} from './dom';
import { renderToTemplate } from './template';
import { AbstractBlock } from './types';
import { arrayMount$, arrayPatch$ } from './array';
import {
  TEXT_NODE_CACHE,
  AttributeFlag,
  ChildFlag,
  EventFlag,
  StyleAttributeFlag,
  EVENT_PATCH,
} from './constants';
import type { ArrayBlock } from './array';
import type { EditChild, VElement, Hole, VNode, Edit } from './types';

const HOLE_PROXY = new Proxy(
  {},
  {
    // A universal getter will return a Hole instance if props[any] is accessed
    // Allows code to identify holes in virtual nodes ("digs" them out)
    get(_, key: string): Hole {
      return { $: key };
    },
  }
);

export const block = (
  fn: (props?: MillionProps) => VElement,
  unwrap?: (vnode: VElement) => VNode,
  shouldUpdate?: (oldProps: MillionProps, newProps: MillionProps) => boolean,
  svg?: boolean,
) => {
  const vnode = fn(HOLE_PROXY);
  const edits: Edit[] = [];

  // Turns vnode into a string of HTML and creates an array of "edits"
  // Edits are instructions for how to update the DOM given some props
  const root = stringToDOM(
    renderToTemplate(unwrap ? unwrap(vnode) : vnode, edits)
  );

  return <T extends MillionProps>(
    props?: T | null,
    key?: string,
    shouldUpdateCurrentBlock?: (
      oldProps: MillionProps,
      newProps: MillionProps
    ) => boolean
  ) => {
    return new Block(
      root,
      edits,
      props,
      key ?? props?.key ?? null,
      shouldUpdateCurrentBlock ?? shouldUpdate ?? null,
      null
    );
  };
};

export const mount = (
  block: AbstractBlock,
  parent?: HTMLElement,
  hydrateNode?: HTMLElement
): Node => {
  if ('b' in block && parent) {
    return arrayMount$.call(block, parent, null);
  }
  return mount$.call(block, parent, null, hydrateNode);
};

export const patch = (
  oldBlock: AbstractBlock,
  newBlock: AbstractBlock
): Node => {
  if ('b' in oldBlock || 'b' in newBlock) {
    arrayPatch$.call(oldBlock, newBlock as ArrayBlock);
  }

  if (!oldBlock.l) mount$.call(oldBlock, undefined, null, null);
  if ((oldBlock.k && oldBlock.k === newBlock.k) || oldBlock.r === newBlock.r) {
    return patch$.call(oldBlock, newBlock);
  }
  const el = mount$.call(newBlock, oldBlock.t()!, oldBlock.l, null);
  remove$.call(oldBlock);
  oldBlock.k = newBlock.k!;
  return el;
};

export class Block extends AbstractBlock {
  declare r: Node;
  declare e: Edit[];
  declare parentEl: HTMLElement;
  declare boundaries: [Node, Node];

  constructor(
    root: Node,
    edits: Edit[],
    props?: MillionProps | null,
    key?: string | null,
    shouldUpdate?:
      | ((oldProps: MillionProps, newProps: MillionProps) => boolean)
      | null,
    getElements?: ((root: Node) => HTMLElement[]) | null
  ) {
    super();
    this.r = root;
    this.d = props;
    this.e = edits;
    this.k = key;
    this.c = Array(edits.length);
    if (shouldUpdate) {
      this._u = shouldUpdate;
    } else {
      this._u = null;
    }
    if (getElements) {
      this.g = getElements;
    } else {
      this.g = null;
    }

    this.boundaries = [
      document.createComment('start'),
      document.createComment('end'),
    ];
  }
  m(
    parentEl?: HTMLElement,
    refNode: Node | null = null,
    hydrateNode?: HTMLElement | null
  ): Node {
    this.parentEl = parentEl;
    if (this.l) return this.l;
    // cloneNode(true) uses less memory than recursively creating new nodes
    const root = hydrateNode ?? cloneNode$.call(this.r, true);
    const elements = this.g?.(root);
    if (elements) this.c = elements;

    for (let i = 0, j = this.e.length; i < j; ++i) {
      const current = this.e[i]!;
      const el = (elements?.[i] ??
        getCurrentElement(current.p!, root, this.c, i)) as HTMLElement;
      removeComments(el);

      for (let k = 0, l = current.e.length; k < l; ++k) {
        const edit = current.e[k]!;
        const value = this.d![edit.h];

        if (edit.t & ChildFlag) {
          if (value instanceof AbstractBlock) {
            const child = childAt(el, edit.i!);
            if (hydrateNode) {
              value.m(el, child, child);
            } else {
              value.m(el, child);
            }
            continue;
          }
          if (!el[TEXT_NODE_CACHE]) el[TEXT_NODE_CACHE] = new Array(l);

          if (value && typeof value === 'object' && 'foreign' in value) {
            const targetEl = value.current;
            el[TEXT_NODE_CACHE][k] = targetEl;

            // insertBefore$.call(el, targetEl , childAt(el, edit.i))
            if (value.p) {
              insertBefore$.call(
                el,
                value.p.commentMarker,
                childAt(el, edit.i)
              );
              value.p.parent = el;
              value.p.pingResolve();
              value.p.pongPromise.then(() => {
                // debugger
                const parentOfComment = (value.p.commentMarker as Node)
                  .parentNode!;
                const i = Array.prototype.indexOf.call(
                  parentOfComment.childNodes,
                  value.p.commentMarker
                );
                if (parentOfComment !== targetEl) {
                  insertBefore$.call(
                    parentOfComment,
                    targetEl,
                    childAt(parentOfComment, i)
                  );
                }
                value.p.commentMarker.remove();
                // console.log('mounted finish', value.p.commentMarker, childAt(el, edit.i))
                // replaceChild$.call(el, value.p.commentMarker , targetEl);
              });
            }
            continue;
          }
          if (hydrateNode) {
            el[TEXT_NODE_CACHE][k] = childAt(el, edit.i!);
            continue;
          }
          // insertText() on mount, setText() on patch
          el[TEXT_NODE_CACHE][k] = insertText(
            el,
            // eslint-disable-next-line eqeqeq
            value == null || value === false ? '' : String(value),
            edit.i!
          );
        } else if (edit.t & EventFlag) {
          const patch = createEventListener(el, edit.n!, value);
          el[EVENT_PATCH + edit.n!] = patch;
        } else if (edit.t & AttributeFlag) {
          setAttribute(el, edit.n!, value);
        } else if (edit.t & StyleAttributeFlag) {
          if (typeof value === 'string' || typeof value === 'number') {
            setStyleAttribute(el, edit.n!, value);
          } else {
            for (const style in value) {
              setStyleAttribute(el, style, value[style]);
            }
          }
        } else {
          setSvgAttribute(el, edit.n!, value);
        }
      }

      const initsLength = current.i?.length;
      if (!initsLength) continue;
      for (let k = 0; k < initsLength; ++k) {
        const init = current.i![k]!;

        if (init.t & ChildFlag) {
          // Handles case for positioning text nodes. When text nodes are
          // put into a template, they can be merged. For example,
          // ["hello", "world"] becomes "helloworld" in the DOM.
          // Inserts text nodes into the DOM at the correct position.
          if (init.v && !hydrateNode) insertText(el, init.v, init.i);
        } else if (init.t & EventFlag) {
          createEventListener(el, init.n!, init.l!);
        } else {
          const child = childAt(el, init.i!);
          if (hydrateNode) {
            init.b!.m(el, child, child);
          } else {
            init.b!.m(el, child);
          }
        }
      }
    }

    insertBefore$.call(root, this.boundaries[0], childAt(root, 0));
    appendChild$.call(root, this.boundaries[1]);

    if (parentEl && !hydrateNode) {
      insertBefore$.call(parentEl, root, refNode);
    }
    this.l = root;

    return root;
  }
  p(newBlock: AbstractBlock): Node {
    const root = this.l!;
    if (!newBlock.d) return root;
    const props = this.d!;
    // If props are the same, no need to patch
    if (!shouldUpdate$.call(this, props, newBlock.d)) return root;
    this.d = newBlock.d;

    for (let i = 0, j = this.e.length; i < j; ++i) {
      const current = this.e[i]!;
      const el = (this.c![i] ??
        getCurrentElement(current.p!, root, this.c, i)) as HTMLElement;
      for (let k = 0, l = current.e.length; k < l; ++k) {
        const edit = current.e[k]!;
        const oldValue = props[edit.h];
        const newValue = newBlock.d[edit.h];

        if (newValue === oldValue) continue;

        if (edit.t & EventFlag) {
          el[EVENT_PATCH + edit.n!]!(newValue);
          continue;
        }
        if (edit.t & ChildFlag) {
          if (oldValue instanceof AbstractBlock) {
            // Remember! If we find a block inside a child, we need to locate
            // the cooresponding block in the new props and patch it.
            const firstEdit = newBlock.e?.[i]?.e[k] as EditChild;
            const newChildBlock = newBlock.d[firstEdit.h];
            oldValue.p(newChildBlock);
            continue;
          }
          if (
            newValue &&
            typeof newValue === 'object' &&
            'foreign' in newValue
          ) {
            const targetEl = el[TEXT_NODE_CACHE][k];
            if (newValue.unstable && oldValue !== newValue) {
              const newTargetEl = newValue.current;
              el[TEXT_NODE_CACHE][k] = newTargetEl;
              replaceChild$.call(el, newTargetEl, targetEl);
            } else {
              // debugger
              newValue.current = targetEl;
              appendChild$.call(el, newValue.current);
            }
            // appendChild$.call(this.parentEl, el)
            // newValue.rerender()

            // if (newValue.p) {
            //   newValue.p.parent = el
            //   newValue.p.resolve()
            // }

            continue;
          }
          setText(
            el[TEXT_NODE_CACHE][k],
            // eslint-disable-next-line eqeqeq
            newValue == null || newValue === false ? '' : String(newValue)
          );
        } else if (edit.t & AttributeFlag) {
          setAttribute(el, edit.n!, newValue);
        } else if (edit.t & StyleAttributeFlag) {
          if (typeof newValue === 'string' || typeof newValue === 'number') {
            setStyleAttribute(el, edit.n!, newValue);
          } else {
            for (const style in newValue) {
              if (newValue[style] !== oldValue[style]) {
                setStyleAttribute(el, style, newValue[style]);
              }
            }
          }
        } else {
          setSvgAttribute(el, edit.n!, newValue);
        }
      }
    }

    if (root instanceof DocumentFragment) {
      // const endComment = this.boundaries[1]
      // const parentOfComment =endComment.parentNode!
      // const i = Array.prototype.indexOf.call(parentOfComment.childNodes, endComment)
      // insertBefore$.call(parentOfComment, root , childAt(parentOfComment, i))
      appendChild$.call(this.parentEl, root);
    }

    return root;
  }
  v(block: AbstractBlock | null = null, refNode: Node | null = null): void {
    insertBefore$.call(this.t(), this.l!, block ? block.l! : refNode);
  }
  x(): void {
    betweenNodes(this.boundaries[0], this.boundaries[1]).forEach((n) =>
      n.parentNode?.removeChild(n)
    );
    this.l?.parentNode?.removeChild(this.l)
    this.boundaries.map((n) => n.parentNode?.removeChild(n))
    this.l = null;
  }
  u(_oldProps: MillionProps, _newProps: MillionProps): boolean {
    if (!this._u) return true;
    return this._u(_oldProps, _newProps);
  }
  s(): string {
    return String((this.l as HTMLElement).outerHTML);
  }
  t(): HTMLElement | null | undefined {
    if (!this._t) this._t = this.l?.parentElement;
    return this._t;
  }
}

const getCurrentElement = (
  path: number[],
  root: Node,
  cache?: Node[],
  key?: number
): Node => {
  const pathLength = path.length;
  if (!pathLength) return root;
  const isCacheAndKeyExists = cache && key !== undefined;
  if (isCacheAndKeyExists && cache[key]) {
    return cache[key]!;
  }
  // path is an array of indices to traverse the DOM tree
  // For example, [0, 1, 2] becomes:
  // root.firstChild.firstChild.nextSibling.firstChild.nextSibling.nextSibling
  // We use path because we don't have the actual DOM nodes until mount()
  for (let i = 0; i < pathLength; ++i) {
    const siblings = path[i]!;
    root = childAt(root, siblings);
  }

  if (isCacheAndKeyExists) cache[key] = root;
  return root;
};

export const withKey = (value: any, key: string): any => {
  value.key = key;
  return value;
};

const block$ = Block.prototype;

export const mount$ = block$.m;
export const patch$ = block$.p;
export const move$ = block$.v;
export const remove$ = block$.x;
export const shouldUpdate$ = block$.u;
