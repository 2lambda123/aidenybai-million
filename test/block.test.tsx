// eslint-disable-next-line eslint-comments/disable-enable-pair
/* eslint-disable @typescript-eslint/restrict-template-expressions */
// eslint-disable-next-line eslint-comments/disable-enable-pair
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { describe, it } from 'vitest';
import { block as createBlock } from '../packages/million';
import type { Props, VElement } from '../packages/million';

const fn = (props?: Props): VElement => (
  <div>
    <h1>Hello</h1> World
    <p title="baz" className={props?.bar}>
      {props?.foo}
    </p>
  </div>
);

const styleFn = (props?: Props): VElement => (
  <div>
    <h1 className={`${props?.bar}_color`}>Hello</h1> World
    <p style={{ [props?.foo]: props?.bar }}>{`${props?.zoo}abc`}</p>
  </div>
);

describe.concurrent('block', () => {
  it('should create reusable block', ({ expect }) => {
    const block = createBlock(fn);
    expect(block).toBeDefined();
    expect(block({ foo: 'foo', bar: 'bar' })).toBeDefined();
  });
  it('should mount block', ({ expect }) => {
    const block = createBlock(fn);
    const main = block({ foo: 'foo', bar: 'bar' });
    main.m();
    expect(main.l?.outerHTML).toEqual(
      '<div><h1>Hello</h1> World<p title="baz" class="bar">foo</p></div>',
    );
  });
  it('should patch block', ({ expect }) => {
    const block = createBlock(fn);
    const main = block({ foo: 'foo', bar: 'bar' });
    main.m();
    main.p(block({ foo: 'bar', bar: 'foo' }));
    expect(main.l?.outerHTML).toEqual(
      '<div><h1>Hello</h1> World<p title="baz" class="foo">bar</p></div>',
    );
    main.p(block({ foo: 'bar', bar: 'bar' }));
    expect(main.l?.outerHTML).toEqual(
      '<div><h1>Hello</h1> World<p title="baz" class="bar">bar</p></div>',
    );
  });
  it('should patch nested blocks', ({ expect }) => {
    const block = createBlock(fn);
    const subBlock = createBlock(fn);
    const main = block({ foo: subBlock({ foo: '1', bar: '2' }), bar: 'bar' });
    main.m();
    main.p(block({ foo: subBlock({ foo: '2', bar: '1' }), bar: 'bar' }));
    expect(main.l?.outerHTML).toEqual(
      '<div><h1>Hello</h1> World<p title="baz" class="bar"><div><h1>Hello</h1> World<p title="baz" class="1">2</p></div></p></div>',
    );
  });
  it('should remove block', ({ expect }) => {
    const block = createBlock(fn);
    const main = block({ foo: 'foo', bar: 'bar' });
    main.m();
    main.x();
    expect(main.l).toBeNull();
  });
  it('should ignore null, undefined, false', ({ expect }) => {
    const block = createBlock(fn);
    const main = block({ foo: null, bar: 'bar' });
    main.m();
    expect(main.l?.outerHTML).toEqual(
      '<div><h1>Hello</h1> World<p title="baz" class="bar"></p></div>',
    );
    main.p(block({ foo: undefined, bar: 'bar' }));
    expect(main.l?.outerHTML).toEqual(
      '<div><h1>Hello</h1> World<p title="baz" class="bar"></p></div>',
    );
    main.p(block({ foo: false, bar: 'bar' }));
    expect(main.l?.outerHTML).toEqual(
      '<div><h1>Hello</h1> World<p title="baz" class="bar"></p></div>',
    );
  });
  it('should patch style', ({ expect }) => {
    const block = createBlock(styleFn);
    const main = block({ foo: 'backgroundColor', bar: 'red', zoo: 'zoo' });
    main.m();
    expect(main.l?.outerHTML).toEqual(
      '<div><h1 class="red_color">Hello</h1> World<p style="background-color: red;">zooabc</p></div>',
    );
    main.p(block({ foo: 'background', bar: 'blue', zoo: 'bazz' }));
    expect(main.l?.outerHTML).toEqual(
      '<div><h1 class="blue_color">Hello</h1> World<p style="background: blue;">bazzabc</p></div>',
    );
  });
});
