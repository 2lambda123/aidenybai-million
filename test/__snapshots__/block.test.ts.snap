// Vitest Snapshot v1

exports[`block() > should transforms a valid block correctly 1`] = `
"import { block } from \\"million/react\\";
function Lion() {
  return /*#__PURE__*/React.createElement(\\"img\\", {
    src: \\"https://million.dev/lion.svg\\"
  });
}
const LionBlock = block(Lion);"
`;
